File: apps/android/feature/library/src/main/kotlin/app/storyarc/feature/library/LibraryFilterMenu.kt
package app.storyarc.feature.library

import androidx.compose.foundation.layout.Box
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Checkbox
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.FilterChip
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.res.pluralStringResource
import androidx.compose.ui.res.stringResource
import app.storyarc.core.model.LibraryQuery
import app.storyarc.core.model.LibraryScope
import app.storyarc.core.model.ReadState
import app.storyarc.core.model.SourceRegistry
import app.storyarc.core.model.YearRange
import java.util.Locale

/**
 * One group of alternatives the reader can narrow by.
 *
 * `library-browsing` names ten facets; these are the nine the app can answer. The
 * enum exists so the menu can show one group at a time — see [FilterChipMenu].
 *
 * [LIBRARY] leads because it is the newest and the one that changed shape: narrowing to a
 * single library used to be a *scope* — a mode with its own control in the top bar, which
 * silently narrowed the search as well and which a reader could be left in without
 * noticing. It is a filter now, cleared by the same action that clears every other filter
 * and counted in the same badge.
 *
 * [DOWNLOAD] sits with [READ_STATE] because the two are the reader's own relationship with a
 * book rather than anything a file says about itself. It is not the availability chip in the
 * row above wearing a second name — [DownloadFilter] sets out the difference at length.
 */
private enum class FilterSection {
    LIBRARY,
    READ_STATE,
    DOWNLOAD,
    FORMAT,
    LANGUAGE,
    PUBLISHER,
    GENRE,
    TAG,
    DECADE,
}

/**
 * What the library is narrowed to.
 *
 * `library-browsing`: the groups combine with AND, the active count is visible on
 * the control, and one action clears them all. The menu shows one group at a time —
 * a flat menu listing every publisher, genre and tag a real library holds would run
 * past the bottom of the screen long before the reader reached "Clear filters", and
 * the reader would have to scroll a menu to undo a mistake.
 *
 * A chip rather than an icon button. The chip says how many filters are active in a word
 * a reader can read at a glance, which the tinted funnel glyph it replaces could not — and
 * it sits in [LibraryControls] beside the axis and the sort, where the three narrowing
 * decisions belong together.
 */
@Composable
internal fun FilterChipMenu(
    query: LibraryQuery,
    registry: SourceRegistry,
    downloads: DownloadFilter,
    viewModel: LibraryViewModel,
    onQueryChange: (LibraryQuery) -> Unit,
    onDownloadsChange: (DownloadFilter) -> Unit,
    onClearFilters: () -> Unit,
) {
    var open by remember { mutableStateOf(false) }
    var section by remember { mutableStateOf<FilterSection?>(null) }
    val active = narrowingCount(query, downloads)

    // The chip and its menu are one item of [LibraryControls]'s wrapping row, not two. A
    // `DropdownMenu` is a popup and measures as nothing, but it still takes a slot -- and a
    // slot of nothing with the row's spacing either side of it is a gap that can push the
    // next chip onto a line it did not need.
    Box {
        FilterChip(
            selected = active > 0,
            onClick = { open = true },
            label = {
                Text(
                    text = if (active > 0) {
                        // A plural, not a format. "1 filters active" is wrong in every
                        // language, and the count reaches 1 whenever a reader sets one
                        // filter.
                        pluralStringResource(R.plurals.library_filter_active, active, active)
                    } else {
                        stringResource(R.string.library_filter)
                    },
                )
            },
        )
        DropdownMenu(
            expanded = open,
            onDismissRequest = {
                open = false
                // Reopening lands on the group list rather than wherever the reader was
                // three taps ago, which they have no reason to remember.
                section = null
            },
        ) {
            when (val chosen = section) {
                null -> SectionList(
                    query = query,
                    registry = registry,
                    downloads = downloads,
                    viewModel = viewModel,
                    onOpen = { section = it },
                    onClear = {
                        onClearFilters()
                        open = false
                    },
                )

                else -> {
                    BackItem(chosen) { section = null }
                    SectionValues(
                        chosen,
                        query,
                        registry,
                        downloads,
                        viewModel,
                        onQueryChange,
                        onDownloadsChange,
                    )
                }
            }
        }
    }
}

/**
 * How much of the view the reader has narrowed, the library filter and the download group
 * included.
 *
 * `LibraryQuery.activeFilterCount` counts the seven facets it holds and cannot count the
 * other two, which are fields beside it rather than in it. Counted here so the badge matches
 * what "Clear filters" undoes — a chip reading "2 filters active" that clears three things is
 * a chip nobody trusts twice.
 */
private fun narrowingCount(query: LibraryQuery, downloads: DownloadFilter): Int =
    query.activeFilterCount +
        (if (query.scope == LibraryScope.AllSources) 0 else 1) +
        (if (downloads.isActive) 1 else 0)

/** The groups themselves, each one worth opening only if the library has values for it. */
@Composable
private fun SectionList(
    query: LibraryQuery,
    registry: SourceRegistry,
    downloads: DownloadFilter,
    viewModel: LibraryViewModel,
    onOpen: (FilterSection) -> Unit,
    onClear: () -> Unit,
) {
    FilterSection.entries.forEach { section ->
        // A group with nothing in it is left out entirely: an empty "Genre" list
        // tells the reader nothing and costs a tap to find out.
        if (section.hasValues(registry, viewModel)) {
            SectionItem(
                label = stringResource(section.labelRes),
                isActive = section.isActive(query, downloads),
                onClick = { onOpen(section) },
            )
        }
    }
    if (narrowingCount(query, downloads) > 0) {
        HorizontalDivider()
        DropdownMenuItem(
            text = { Text(stringResource(R.string.library_filter_clear)) },
            onClick = onClear,
        )
    }
}

/** The values of one group, as the reader ticks them. */
@Composable
private fun SectionValues(
    section: FilterSection,
    query: LibraryQuery,
    registry: SourceRegistry,
    downloads: DownloadFilter,
    viewModel: LibraryViewModel,
    onQueryChange: (LibraryQuery) -> Unit,
    onDownloadsChange: (DownloadFilter) -> Unit,
) {
    when (section) {
        FilterSection.LIBRARY -> LibraryValues(query, registry, onQueryChange)

        FilterSection.READ_STATE -> ReadState.entries.forEach { state ->
            CheckedItem(stringResource(state.labelRes), state in query.readStates) {
                onQueryChange(query.copy(readStates = toggled(query.readStates, state)))
            }
        }

        FilterSection.DOWNLOAD -> DownloadValues(downloads, onDownloadsChange)

        FilterSection.FORMAT -> viewModel.availableFormats().forEach { format ->
            CheckedItem(format.displayName, format in query.formats) {
                onQueryChange(query.copy(formats = toggled(query.formats, format)))
            }
        }

        // Languages named in themselves. A reader looking for Deutsch is not helped
        // by "German", which is the rule the language setting already follows.
        FilterSection.LANGUAGE -> viewModel.availableLanguages().forEach { code ->
            CheckedItem(languageName(code), code in query.languages) {
                onQueryChange(query.copy(languages = toggled(query.languages, code)))
            }
        }

        FilterSection.PUBLISHER -> viewModel.availablePublishers().forEach { publisher ->
            CheckedItem(publisher, publisher in query.publishers) {
                onQueryChange(query.copy(publishers = toggled(query.publishers, publisher)))
            }
        }

        FilterSection.GENRE -> viewModel.availableGenres().forEach { genre ->
            CheckedItem(genre, genre in query.genres) {
                onQueryChange(query.copy(genres = toggled(query.genres, genre)))
            }
        }

        FilterSection.TAG -> viewModel.availableTags().forEach { tag ->
            CheckedItem(tag, tag in query.tags) {
                onQueryChange(query.copy(tags = toggled(query.tags, tag)))
            }
        }

        FilterSection.DECADE -> DecadeValues(query, viewModel, onQueryChange)
    }
}

/**
 * The libraries a reader has added, offered by name.
 *
 * Radio buttons rather than checkboxes: a publication comes from one library, so a set of
 * them is a question the shelf cannot answer. "Everywhere" is how the filter is turned back
 * off — the same act as unticking the last value in any other group.
 *
 * The registry's order, because `sources` makes that order meaningful and a list that
 * reshuffled it would undo an arrangement the reader made by hand.
 */
@Composable
private fun LibraryValues(
    query: LibraryQuery,
    registry: SourceRegistry,
    onQueryChange: (LibraryQuery) -> Unit,
) {
    ChosenItem(
        label = stringResource(R.string.library_scope_all),
        chosen = query.scope == LibraryScope.AllSources,
    ) {
        onQueryChange(query.copy(scope = LibraryScope.AllSources))
    }
    registry.sources.forEach { source ->
        val scope = LibraryScope.OneSource(source.id)
        ChosenItem(label = source.displayName, chosen = query.scope == scope) {
            onQueryChange(query.copy(scope = scope))
        }
    }
}

/**
 * Whether the app fetched it, per `library-browsing`'s *Filtering offline*.
 *
 * Radio buttons rather than the checkboxes most groups use, and for the reason the decade
 * group has them: a publication is downloaded or it is not, so ticking both answers is the
 * same as ticking neither. "Downloaded or not" is how the group is turned back off, and it is
 * the group's own name because that is exactly what it shows.
 */
@Composable
private fun DownloadValues(downloads: DownloadFilter, onChange: (DownloadFilter) -> Unit) {
    listOf(
        DownloadFilter.EITHER to R.string.library_filter_download,
        DownloadFilter.DOWNLOADED to R.string.library_filter_download_yes,
        DownloadFilter.NOT_DOWNLOADED to R.string.library_filter_download_no,
    ).forEach { (value, label) ->
        ChosenItem(stringResource(label), downloads == value) { onChange(value) }
    }
}

/**
 * The year range, offered as the decades the library actually spans.
 *
 * Radio buttons rather than the checkboxes most groups use, because a range
 * is one answer and not a set of them. "Any year" is how it is turned back off — the
 * same act as unticking the last value in any other group.
 */
@Composable
private fun DecadeValues(
    query: LibraryQuery,
    viewModel: LibraryViewModel,
    onQueryChange: (LibraryQuery) -> Unit,
) {
    ChosenItem(stringResource(R.string.library_filter_decade_any), !query.years.isActive) {
        onQueryChange(query.copy(years = YearRange()))
    }
    viewModel.availableDecades().forEach { start ->
        ChosenItem(
            label = stringResource(R.string.library_filter_decade_label, start),
            // Not `years == YearRange(start, start + 9)`: a range set to something
            // that is not a decade is one this control cannot draw, and showing
            // nothing selected is the honest answer to that.
            chosen = query.years.from == start,
        ) {
            onQueryChange(query.copy(years = YearRange(from = start, to = start + 9)))
        }
    }
}

/** A group, and whether the reader has set any of it. */
@Composable
private fun SectionItem(label: String, isActive: Boolean, onClick: () -> Unit) {
    DropdownMenuItem(
        text = { Text(label) },
        // The tick is the only thing that says a collapsed group is narrowing the
        // view. Without it the badge would report three active filters and the menu
        // would look untouched.
        leadingIcon = if (isActive) {
            { Icon(Icons.Filled.Check, contentDescription = null) }
        } else {
            null
        },
        trailingIcon = {
            Icon(Icons.AutoMirrored.Filled.KeyboardArrowRight, contentDescription = null)
        },
        onClick = onClick,
    )
}

/** The way back to the group list, which also names where the reader is. */
@Composable
private fun BackItem(section: FilterSection, onBack: () -> Unit) {
    DropdownMenuItem(
        text = { Text(stringResource(section.labelRes)) },
        leadingIcon = {
            Icon(
                imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                contentDescription = stringResource(R.string.library_filter_back),
            )
        },
        onClick = onBack,
    )
    HorizontalDivider()
}

@Composable
private fun CheckedItem(label: String, checked: Boolean, onToggle: () -> Unit) {
    DropdownMenuItem(
        text = { Text(label) },
        leadingIcon = { Checkbox(checked = checked, onCheckedChange = null) },
        onClick = onToggle,
    )
}

@Composable
private fun ChosenItem(label: String, chosen: Boolean, onChoose: () -> Unit) {
    DropdownMenuItem(
        text = { Text(label) },
        leadingIcon = { RadioButton(selected = chosen, onClick = null) },
        onClick = onChoose,
    )
}

private fun <T> toggled(current: Set<T>, value: T): Set<T> =
    if (value in current) current - value else current + value

/**
 * A language code as its own speakers write it.
 *
 * The same rule Settings follows, duplicated rather than shared: a feature module
 * never depends on another feature module, and this is four lines.
 */
private fun languageName(code: String): String {
    val locale = Locale.forLanguageTag(code)
    return locale.getDisplayLanguage(locale)
        .ifEmpty { code }
        .replaceFirstChar { it.titlecase(locale) }
}

/**
 * How the filter groups are named on screen.
 *
 * The enums live in `:core:model` and carry no resources: the domain has no
 * business holding UI copy. Naming them is presentation, so it lives here.
 */
private val FilterSection.labelRes: Int
    get() = when (this) {
        FilterSection.LIBRARY -> R.string.library_filter_library
        FilterSection.READ_STATE -> R.string.library_filter_read_state
        FilterSection.DOWNLOAD -> R.string.library_filter_download
        FilterSection.FORMAT -> R.string.library_filter_format
        FilterSection.LANGUAGE -> R.string.library_filter_language
        FilterSection.PUBLISHER -> R.string.library_filter_publisher
        FilterSection.GENRE -> R.string.library_filter_genre
        FilterSection.TAG -> R.string.library_filter_tag
        FilterSection.DECADE -> R.string.library_filter_decade
    }

private fun FilterSection.isActive(
    query: LibraryQuery,
    downloads: DownloadFilter,
): Boolean = when (this) {
    FilterSection.LIBRARY -> query.scope != LibraryScope.AllSources
    FilterSection.READ_STATE -> query.readStates.isNotEmpty()
    FilterSection.DOWNLOAD -> downloads.isActive
    FilterSection.FORMAT -> query.formats.isNotEmpty()
    FilterSection.LANGUAGE -> query.languages.isNotEmpty()
    FilterSection.PUBLISHER -> query.publishers.isNotEmpty()
    FilterSection.GENRE -> query.genres.isNotEmpty()
    FilterSection.TAG -> query.tags.isNotEmpty()
    FilterSection.DECADE -> query.years.isActive
}

private fun FilterSection.hasValues(
    registry: SourceRegistry,
    viewModel: LibraryViewModel,
): Boolean = when (this) {
    // Only with a second library to narrow to. A group whose whole list is "Everywhere"
    // and the one library there is asks the reader nothing.
    FilterSection.LIBRARY -> registry.sources.size > 1
    // Always offered: the three read states exist whether or not anything is in them,
    // and "Unread" over an empty result is an answer rather than a dead end.
    FilterSection.READ_STATE -> true
    // Always offered too, and for the same reason. A library with nothing downloaded still
    // answers "Not downloaded" usefully — that is the question asked the night before a
    // journey, not during one.
    FilterSection.DOWNLOAD -> true
    FilterSection.FORMAT -> viewModel.availableFormats().isNotEmpty()
    FilterSection.LANGUAGE -> viewModel.availableLanguages().isNotEmpty()
    FilterSection.PUBLISHER -> viewModel.availablePublishers().isNotEmpty()
    FilterSection.GENRE -> viewModel.availableGenres().isNotEmpty()
    FilterSection.TAG -> viewModel.availableTags().isNotEmpty()
    FilterSection.DECADE -> viewModel.availableDecades().isNotEmpty()
}

private val ReadState.labelRes: Int
    get() = when (this) {
        ReadState.UNREAD -> R.string.library_read_state_unread
        ReadState.IN_PROGRESS -> R.string.library_read_state_in_progress
        ReadState.FINISHED -> R.string.library_read_state_finished
    }

File: apps/ios/Packages/StoryArcKit/Sources/LibraryFeature/LibraryFilterMenu.swift
internal import SwiftUI

internal import DesignSystem
internal import StoryArcCore

/// What the library is narrowed to.
///
/// `library-browsing`: the groups combine with AND, the active count is visible on
/// the control, and one action clears them all. None of that changed when the
/// groups went from two to seven — what changed is that each group is now a
/// submenu. A flat menu listing every publisher, genre and tag a real library holds
/// would run past the bottom of the screen long before the reader reached "Clear
/// filters", and the reader would have to scroll a menu to undo a mistake.
///
/// Split out of `LibraryBrowsingControls` for the same reason it grew: the menu is
/// now longer than the two controls it used to sit beside.
struct FilterMenu: View {
    let model: LibraryModel

    /// The download group, which is not on the query.
    ///
    /// ``DownloadFilter`` says why it is not: the query is the value both platforms encode,
    /// and a case added to it is a change to `StoryArcCore` and to Android's mirror. So the
    /// screen owns it and this menu is handed it, the way ``ScopeMenu`` is handed the
    /// availability axis.
    @Binding var downloads: DownloadFilter

    /// The primary axis, cleared with the filters and counted with none of them.
    ///
    /// See ``LibraryNarrowing/activeCount``: availability has a control of its own that states it
    /// while it is set, so it is not part of the badge — and *Clear filters* still puts it
    /// back, because a cleared shelf that is still showing only what is on this device is a
    /// shelf as empty as the reader found it.
    @Binding var availability: LibraryAvailability

    var body: some View {
        Menu {
            // First, as on Android, and for the reason its own `FilterSection` gives:
            // narrowing to one library is the newest of these groups and the one that
            // changed shape. It was a scope with a control of its own in the toolbar, which
            // silently narrowed the search as well and which a reader could be left in
            // without noticing. It is a filter now — cleared by the action that clears every
            // other filter, and counted in the same badge.
            libraries

            group("library.filter.readState", isActive: !model.query.readStates.isEmpty) {
                ForEach(ReadState.allCases, id: \.self) { state in
                    Toggle(isOn: readState(state)) {
                        Text(state.titleKey, bundle: .module)
                    }
                }
            }

            downloadState
            formats
            values("library.filter.language", languageNames, \.languages)
            values("library.filter.publisher", pairs(model.availablePublishers), \.publishers)
            values("library.filter.genre", pairs(model.availableGenres), \.genres)
            values("library.filter.tag", pairs(model.availableTags), \.tags)
            decades

            if narrowing.isActive {
                Divider()
                Button(role: .destructive, action: clear) {
                    Text("library.filter.clear", bundle: .module)
                }
            }
        } label: {
            // **The count is drawn now, and it used to be spoken only.** The comment this
            // replaced said a menu label "cannot carry" a badge, which is true of
            // `.badge(_:)` and not of the label's own content: a glyph and a number beside it
            // is a label like any other. So a reader who could not hear VoiceOver knew *that*
            // the shelf was narrowed — the funnel fills — and never by how much, on a screen
            // whose whole job is to show them a set they have narrowed. `design.md` forbids a
            // state carried by appearance alone, and one filter looked exactly like six.
            HStack(spacing: StoryArcSpace.hair) {
                Image(
                    systemName: narrowing.isActive
                        ? "line.3.horizontal.decrease.circle.fill"
                        : "line.3.horizontal.decrease.circle"
                )
                if narrowing.isActive {
                    // Monospaced digits so the toolbar item does not resize as the count
                    // crosses from one glyph width to another while a reader is using it.
                    Text(narrowing.activeCount, format: .number)
                        .monospacedDigit()
                }
            }
            // The words, for a reader who gets neither the glyph nor the digit.
            .accessibilityLabel(Text("library.filter", bundle: .module))
        }
        .accessibilityValue(
            narrowing.isActive
                ? Text("library.filter.active \(narrowing.activeCount)", bundle: .module)
                : Text(verbatim: "")
        )
    }

    /// Everything this control narrows by, and what it says about it.
    ///
    /// Not `model.query` alone: the library narrowing, the download group and the
    /// availability axis are kept in three places for three good reasons, and until
    /// ``LibraryNarrowing`` existed each call site joined them up for itself — which is how
    /// the count came to omit the one narrowing a reader could not otherwise see.
    private var narrowing: LibraryNarrowing {
        LibraryNarrowing(query: model.query, downloads: downloads, availability: availability)
    }

    /// One action, everything it undoes.
    private func clear() {
        let cleared = narrowing.cleared()
        model.query = cleared.query
        downloads = cleared.downloads
        availability = cleared.availability
    }

    // MARK: - Groups

    /// Which library, when there is more than one to choose between.
    ///
    /// A radio list rather than the checkboxes most groups use: the shelf shows one library
    /// or all of them, so ticking two answers would be the same as ticking neither. "Any
    /// library" is how the group is turned back off, which is the same act as unticking the
    /// last value anywhere else.
    @ViewBuilder
    private var libraries: some View {
        let offered = LibraryNarrowing.offeredLibraries(in: model.registry)
        if !offered.isEmpty {
            group("library.filter.library", isActive: narrowing.isScoped) {
                Picker(selection: scope) {
                    ForEach(offered, id: \.self) { scope in
                        name(of: scope).tag(scope)
                    }
                } label: {
                    Text("library.filter.library", bundle: .module)
                }
            }
        }
    }

    /// What one library is called in the group.
    ///
    /// "Any library" rather than "Everywhere" for the unnarrowed case: the word belongs to
    /// the availability axis in the control next door, and two rows reading "Everywhere" in
    /// one toolbar would be two different promises wearing one name.
    private func name(of scope: LibraryScope) -> Text {
        guard let name = model.registry.name(of: scope.sourceID) else {
            return Text("library.availability.from.all", bundle: .module)
        }
        return Text(name)
    }

    private var scope: Binding<LibraryScope> {
        Binding(get: { model.query.scope }, set: { model.query.scope = $0 })
    }

    /// One group of alternatives, and whether the reader has set any of them.
    ///
    /// The tick on the label is the only thing that says a collapsed group is
    /// narrowing the view. Without it the badge would report three active filters
    /// and the menu would look untouched.
    @ViewBuilder
    private func group(
        _ title: LocalizedStringKey,
        isActive: Bool,
        @ViewBuilder content: () -> some View
    ) -> some View {
        Menu {
            content()
        } label: {
            if isActive {
                Label { Text(title, bundle: .module) } icon: { Image(systemName: "checkmark") }
            } else {
                Text(title, bundle: .module)
            }
        }
    }

    /// Whether the app fetched it, per `library-browsing`'s *Filtering offline*.
    ///
    /// A radio list rather than the toggles most groups use, and for the reason the decade
    /// group is one: a publication is downloaded or it is not, so ticking both answers is
    /// the same as ticking neither. "Downloaded or not" is how the group is turned back
    /// off, and it is the group's own name because that is exactly what it shows.
    ///
    /// Always offered, unlike the groups below it. A library with nothing downloaded still
    /// answers "Not downloaded" usefully — that is the question asked the night before a
    /// journey — and an empty result to "Downloaded" is an answer rather than a dead end.
    @ViewBuilder
    private var downloadState: some View {
        group("library.filter.download", isActive: downloads.isActive) {
            Picker(selection: $downloads) {
                Text("library.filter.download", bundle: .module).tag(DownloadFilter.either)
                Text("library.filter.download.yes", bundle: .module).tag(DownloadFilter.downloaded)
                Text("library.filter.download.no", bundle: .module).tag(DownloadFilter.notDownloaded)
            } label: {
                Text("library.filter.download", bundle: .module)
            }
        }
    }

    /// Formats, named by the domain rather than by a string table: "CBZ" is "CBZ"
    /// in all four languages.
    @ViewBuilder
    private var formats: some View {
        if !model.availableFormats.isEmpty {
            group("library.filter.format", isActive: !model.query.formats.isEmpty) {
                ForEach(model.availableFormats, id: \.self) { value in
                    Toggle(isOn: format(value)) { Text(verbatim: value.displayName) }
                }
            }
        }
    }

    /// A group over values the publications themselves supply.
    ///
    /// Omitted entirely when the library holds none: an empty "Genre" submenu tells
    /// the reader nothing and costs a tap to find out.
    @ViewBuilder
    private func values(
        _ title: LocalizedStringKey,
        _ available: [FilterValue],
        _ facet: WritableKeyPath<LibraryQuery, Set<String>>
    ) -> some View {
        if !available.isEmpty {
            group(title, isActive: !model.query[keyPath: facet].isEmpty) {
                ForEach(available) { value in
                    Toggle(isOn: chosen(value.id, in: facet)) { Text(verbatim: value.label) }
                }
            }
        }
    }

    /// The year range, offered as the decades the library actually spans.
    ///
    /// A radio list rather than the toggles every other group uses, because a range
    /// is one answer and not a set of them. "Any year" is how it is turned back off
    /// — the same act as unticking the last value in any other group.
    @ViewBuilder
    private var decades: some View {
        if !model.availableDecades.isEmpty {
            group("library.filter.decade", isActive: model.query.years.isActive) {
                Picker(selection: decade) {
                    Text("library.filter.decade.any", bundle: .module).tag(Int?.none)
                    ForEach(model.availableDecades, id: \.self) { start in
                        Text("library.filter.decade \(start)", bundle: .module).tag(Int?.some(start))
                    }
                } label: {
                    Text("library.filter.decade", bundle: .module)
                }
            }
        }
    }

    // MARK: - Bindings

    private func readState(_ state: ReadState) -> Binding<Bool> {
        Binding(
            get: { model.query.readStates.contains(state) },
            set: { on in
                if on { model.query.readStates.insert(state) } else { model.query.readStates.remove(state) }
            }
        )
    }

    private func format(_ value: PublicationFormat) -> Binding<Bool> {
        Binding(
            get: { model.query.formats.contains(value) },
            set: { on in
                if on { model.query.formats.insert(value) } else { model.query.formats.remove(value) }
            }
        )
    }

    /// Reads and writes the live query rather than a captured copy: a menu stays
    /// open across several taps, and a set captured when it was built would undo
    /// the tap before last.
    private func chosen(
        _ value: String, in facet: WritableKeyPath<LibraryQuery, Set<String>>
    ) -> Binding<Bool> {
        Binding(
            get: { model.query[keyPath: facet].contains(value) },
            set: { on in
                if on {
                    model.query[keyPath: facet].insert(value)
                } else {
                    model.query[keyPath: facet].remove(value)
                }
            }
        )
    }

    /// The chosen decade, as the year its range starts at.
    ///
    /// `nil` when no range is set — and also when a range was set that is not a
    /// decade, which the query allows and this control cannot draw. Showing nothing
    /// selected is the honest answer to that.
    private var decade: Binding<Int?> {
        Binding(
            get: { model.query.years.from },
            set: { start in
                model.query.years = start.map { YearRange(from: $0, to: $0 + 9) } ?? YearRange()
            }
        )
    }

    // MARK: - Values

    /// Languages named in themselves. A reader looking for Deutsch is not helped by
    /// "German", which is the same rule the language setting follows.
    private var languageNames: [FilterValue] {
        model.availableLanguages.map { FilterValue(id: $0, label: InterfaceLanguage.name(of: $0)) }
    }

    private func pairs(_ values: [String]) -> [FilterValue] {
        values.map { FilterValue(id: $0, label: $0) }
    }
}

/// One value a filter group offers: what it matches, and what it is called.
///
/// The two differ for a language — the query holds "de" and the reader reads
/// "Deutsch" — and are the same string everywhere else.
struct FilterValue: Identifiable, Hashable {
    let id: String
    let label: String
}
